### PR TITLE
Hive: Convert the CREATE TABLE ... PARTITIONED BY to Iceberg identity partitions

### PR DESCRIPTION
**Pull request by pvary**
After consulting with the field folks they convinced me that it would be beneficial to have the first version of conversion in place for creating partitioned Iceberg tables from Hive. They suggested that even in this limited form this feature can boost adoption by allowing to try out Iceberg tables with partitions without changing the actual SQL commands.

### CODE REVIEW

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 1]
```diff
@@ -197,20 +205,32 @@ private static Schema schema(Properties properties, org.apache.hadoop.hive.metas
     if (properties.getProperty(InputFormatConfig.TABLE_SCHEMA) != null) {
       return SchemaParser.fromJson(properties.getProperty(InputFormatConfig.TABLE_SCHEMA));
     } else {
-      return HiveSchemaUtil.convert(hmsTable.getSd().getCols());
+      if (hmsTable.isSetPartitionKeys() && hmsTable.getPartitionKeys().size() != 0) {
```
Comment (marton-bod) [inline mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:208]: nit: can use `!hmsTable.getPartitionKeys().isEmpty()`

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandler.java [hunk 2]
```diff
@@ -666,45 +666,52 @@ public void testCreateTableWithColumnSpecification() throws IOException {
...
+              testTables.locationForCreateTableSQL(TableIdentifier.of("default", "customers")) +
+              " TBLPROPERTIES ('" + InputFormatConfig.PARTITION_SPEC + "'='" +
+              PartitionSpecParser.toJson(spec) + "')");
         }
     );
   }
 
   @Test
-  public void testCreatePartitionedTable() throws IOException {
+  public void testCreateTableWithColumnSpecificationMultilevelPartitioned() {
```
Comment (marton-bod) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandler.java:697]: Can we make `runPartitionedWriteTest` generic enough to reuse it here too? Would be great to have an execution test for multi-level partition as well.

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 3]
```diff
@@ -197,20 +205,32 @@ private static Schema schema(Properties properties, org.apache.hadoop.hive.metas
...
   }
 
   private static PartitionSpec spec(Schema schema, Properties properties,
       org.apache.hadoop.hive.metastore.api.Table hmsTable) {
 
-    Preconditions.checkArgument(hmsTable.getPartitionKeys() == null || hmsTable.getPartitionKeys().isEmpty(),
-        "Partitioned Hive tables are currently not supported");
-
     if (properties.getProperty(InputFormatConfig.PARTITION_SPEC) != null) {
+      Preconditions.checkArgument(hmsTable.getPartitionKeys() == null || hmsTable.getPartitionKeys().isEmpty(),
```
Comment (marton-bod) [inline mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:223]: nit: shall we check `hmsTable.isSetPartitionKeys()` here? in other places we seem to use that instead of the null check

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 1 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java]: Done

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 3 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java]: Done

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 4]
```diff
@@ -197,20 +205,32 @@ private static Schema schema(Properties properties, org.apache.hadoop.hive.metas
...
   }
 
   private static PartitionSpec spec(Schema schema, Properties properties,
       org.apache.hadoop.hive.metastore.api.Table hmsTable) {
 
-    Preconditions.checkArgument(hmsTable.getPartitionKeys() == null || hmsTable.getPartitionKeys().isEmpty(),
-        "Partitioned Hive tables are currently not supported");
-
     if (properties.getProperty(InputFormatConfig.PARTITION_SPEC) != null) {
+      Preconditions.checkArgument(hmsTable.isSetPartitionKeys() || hmsTable.getPartitionKeys().isEmpty(),
```
Comment (marton-bod) [inline mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:223]: shouldn't this be `!hmsTable.isSetPartitionKeys()`?

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandler.java (hunk 2 above)
Comment (pvary) [reply mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandler.java]: done

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 4 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java]: done

**Review by marton-bod — APPROVED** [review]
LGTM, thanks

**Comment by marton-bod** [comment]
@rdblue Could you please take a look at this as well, whenever suitable? :) Thank you!

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 5]
```diff
@@ -197,20 +205,32 @@ private static Schema schema(Properties properties, org.apache.hadoop.hive.metas
     if (properties.getProperty(InputFormatConfig.TABLE_SCHEMA) != null) {
       return SchemaParser.fromJson(properties.getProperty(InputFormatConfig.TABLE_SCHEMA));
     } else {
-      return HiveSchemaUtil.convert(hmsTable.getSd().getCols());
+      if (hmsTable.isSetPartitionKeys() && !hmsTable.getPartitionKeys().isEmpty()) {
+        // Add partitioning columns to the original column list before creating the Iceberg Schema
+        List<FieldSchema> cols = new ArrayList<>(hmsTable.getSd().getCols());
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:210]: Nit: generally prefer `Lists.newArrayList()` to specific class constructors.

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 6]
```diff
@@ -197,20 +205,32 @@ private static Schema schema(Properties properties, org.apache.hadoop.hive.metas
     if (properties.getProperty(InputFormatConfig.TABLE_SCHEMA) != null) {
       return SchemaParser.fromJson(properties.getProperty(InputFormatConfig.TABLE_SCHEMA));
     } else {
-      return HiveSchemaUtil.convert(hmsTable.getSd().getCols());
+      if (hmsTable.isSetPartitionKeys() && !hmsTable.getPartitionKeys().isEmpty()) {
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:208]: Nit: this could be an `else if` to remove a level of indentation.

Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:208]: When would `isSetPartitionKeys()` be true and `getPartitionKeys()` empty?

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 7]
```diff
@@ -197,20 +205,32 @@ private static Schema schema(Properties properties, org.apache.hadoop.hive.metas
...
-        "Partitioned Hive tables are currently not supported");
-
     if (properties.getProperty(InputFormatConfig.PARTITION_SPEC) != null) {
+      Preconditions.checkArgument(!hmsTable.isSetPartitionKeys() || hmsTable.getPartitionKeys().isEmpty(),
+          "Provide only one of the following: Hive partition specification, or the " +
+              InputFormatConfig.PARTITION_SPEC + " property");
       return PartitionSpecParser.fromJson(schema, properties.getProperty(InputFormatConfig.PARTITION_SPEC));
     } else {
-      return PartitionSpec.unpartitioned();
+      if (hmsTable.isSetPartitionKeys() && !hmsTable.getPartitionKeys().isEmpty()) {
```
Comment (rdblue) [inline mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:228]: Nit: could be `else if` case.

File: build.gradle [hunk 8]
```diff
@@ -542,7 +542,7 @@ project(':iceberg-mr') {
 
   test {
     // testJoinTables / testScanTable
-    maxHeapSize '1500m'
+    maxHeapSize '2500m'
```
Comment (rdblue) [inline build.gradle:545]: Why was this needed? Additional tasks because of partitioning?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks @marton-bod for reviewing, and @pvary for working on this!

It looks good to me. I noted a few nits to fix, but I'm also fine merging this if you don't have time to fix them. I'll wait a day or two and then merge if I don't hear back.

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 5 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:210]: Fix in #2029

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 6 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:208]: Fix in #2029

Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:208]: Theoretically it could be set to an empty list. Checked this way to keep on the safe side.

File: mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 7 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:228]: Fix in #2029

File: build.gradle (hunk 8 above)
Comment (pvary) [reply build.gradle:545]: Yeah. The extra tasks eat more memory :(
